File: backend/src/controllers/notification.controller.ts
import type { Request, Response } from "express";
import { notificationService } from "../services/notification.service.js";
import { ApiResponse } from "../utils/api-response.js";
import type { ListNotificationsQuery } from "../validators/notification.validator.js";

export class NotificationController {
  async list(request: Request, response: Response) {
    const query = request.validated?.query as ListNotificationsQuery | undefined;
    const notifications = await notificationService.listNotifications(
      request.user!.id,
      {
        page: query?.page,
        limit: query?.limit,
        unreadOnly: query?.unreadOnly === "true"
      }
    );
    response.json(new ApiResponse(200, notifications, "Notifications retrieved"));
  }

  async unreadCount(request: Request, response: Response) {
    const result = await notificationService.getUnreadCount(request.user!.id);
    response.json(new ApiResponse(200, result, "Unread count retrieved"));
  }

  async markAsRead(request: Request, response: Response) {
    const notification = await notificationService.markAsRead(
      request.params.notificationId as string,
      request.user!.id
    );
    if (!notification) {
      response.status(404).json(new ApiResponse(404, null, "Notification not found"));
      return;
    }
    response.json(new ApiResponse(200, notification, "Notification marked as read"));
  }

  async markAllAsRead(request: Request, response: Response) {
    const result = await notificationService.markAllAsRead(request.user!.id);
    response.json(new ApiResponse(200, result, "All notifications marked as read"));
  }

  async delete(request: Request, response: Response) {
    const notification = await notificationService.deleteNotification(
      request.params.notificationId as string,
      request.user!.id
    );
    if (!notification) {
      response.status(404).json(new ApiResponse(404, null, "Notification not found"));
      return;
    }
    response.json(new ApiResponse(200, null, "Notification deleted"));
  }

  async clearAll(request: Request, response: Response) {
    const result = await notificationService.clearAllNotifications(request.user!.id);
    response.json(new ApiResponse(200, result, "All notifications cleared"));
  }
}

export const notificationController = new NotificationController();

File: backend-java/src/main/java/com/studyconnect/backend/controller/NotificationController.java
package com.studyconnect.backend.controller;

import com.studyconnect.backend.dto.ApiResponse;
import com.studyconnect.backend.service.NotificationService;
import com.studyconnect.backend.util.SecurityUtil;
import org.springframework.security.core.Authentication;
import org.springframework.validation.annotation.Validated;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PatchMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

@RestController
@Validated
@RequestMapping("/api/notifications")
public class NotificationController {

    private final NotificationService notificationService;

    public NotificationController(NotificationService notificationService) {
        this.notificationService = notificationService;
    }

    @GetMapping
    public ApiResponse<?> list(
            @RequestParam(defaultValue = "1") int page,
            @RequestParam(defaultValue = "20") int limit,
            @RequestParam(defaultValue = "false") boolean unreadOnly,
            Authentication authentication) {
        return ApiResponse.success(
                notificationService.listNotifications(SecurityUtil.currentUserId(authentication), page, limit, unreadOnly),
                "Notifications retrieved"
        );
    }

    @GetMapping("/unread-count")
    public ApiResponse<?> unreadCount(Authentication authentication) {
        return ApiResponse.success(
                notificationService.unreadCount(SecurityUtil.currentUserId(authentication)),
                "Unread count retrieved"
        );
    }

    @PatchMapping("/{notificationId}/read")
    public ApiResponse<?> markAsRead(@PathVariable String notificationId, Authentication authentication) {
        return ApiResponse.success(
                notificationService.markAsRead(notificationId, SecurityUtil.currentUserId(authentication)),
                "Notification marked as read"
        );
    }

    @PatchMapping("/read-all")
    public ApiResponse<?> markAllAsRead(Authentication authentication) {
        return ApiResponse.success(
                notificationService.markAllAsRead(SecurityUtil.currentUserId(authentication)),
                "All notifications marked as read"
        );
    }

    @DeleteMapping("/{notificationId}")
    public ApiResponse<?> delete(@PathVariable String notificationId, Authentication authentication) {
        notificationService.deleteNotification(notificationId, SecurityUtil.currentUserId(authentication));
        return ApiResponse.success(null, "Notification deleted");
    }

    @DeleteMapping
    public ApiResponse<?> clearAll(Authentication authentication) {
        return ApiResponse.success(
                notificationService.clearAllNotifications(SecurityUtil.currentUserId(authentication)),
                "All notifications cleared"
        );
    }
}
